# Tipi di dato

In Python tutto è un oggetto e ogni oggetto ha un tipo. Esistono diversi modi per classificare i tipi di dato. Ad esempio, puoi dividerli in scalari, che non possono essere suddivisi ulteriormente (come un numero), e non scalari, che possono essere suddivisi (come una lista di numeri).

Un altro modo per classificarli è in base alla loro origine:
- Tipi di dato integrati (Built-in Data Types)
- Tipi di dato non integrati nell’interprete


## Booleani

I booleani sono un altro tipo di dato fondamentale in Python e consistono solo in due valori: **True** e **False**. I valori booleani vengono usati principalmente quando confrontiamo diversi valori tra loro (ne parleremo meglio più avanti).
In Python, True ha valore 1 mentre False ha valore 0. Possiamo verificare questo fatto convertendoli in interi:

In [1]:
print(int(True))
print(int(False))

1
0


## Tipi numerici

Il tipo di dato più semplice con cui lavorare è la categoria dei “Numeri”, poiché tutti noi abbiamo già una buona familiarità con essi e sappiamo cosa aspettarci quando si parla di numeri!

I due principali sotto-tipi di numeri che vedremo insieme sono:
	•	Interi
	•	Float (numeri decimali)

(Altri due sotto-tipi numerici sono long e complex, che non tratteremo qui)

### Integer

Gli interi sono numeri senza virgola decimale:

In [2]:
1 + 1

2

In [3]:
f = 1 + 1

In [4]:
type(f)

int

In [5]:
type(1 + 1)

int

In [6]:
1256

1256

In [7]:
type(34572765 * 6779_625_645)

int

Puoi facilmente verificare se un numero è un intero o meno. Ti basta usare la funzione incorporata _type( )_ (parleremo meglio delle funzioni incorporate più avanti). Il risultato dovrebbe essere _int_, come mostrato qui sotto:

In [8]:
type(452347)

int

In [9]:
type(1_456_099)

int

### Float

Nel linguaggio Python, i numeri con la virgola decimale sono chiamati Float.

In [10]:
3.14159

3.14159

In [11]:
type(3.14159)

float

### Conversione (cast) tra interi e float

Usando **float( )** e **int( )** possiamo convertire gli interi in float e viceversa:

In [15]:
val = 34.67
type(val)
int(val)
print(val)

34.67


In [16]:
print(int(45.89)) # convertire un float in int
print(int('347')) # convertire una stringa in int

45
347


In [17]:
print(int('56.76')) # questo non funziona

ValueError: invalid literal for int() with base 10: '56.76'

In [18]:
print(int(float('56.76'))) # questo sì...perché?

56


In [19]:
#Lo stesso non vale per la conversione da stringa a float.
#Significa che la conversione funziona solo se la stringa ha la forma di un float.

float('4')

4.0

**IL TUO TURNO**: Considerando a = 3.14, quale codice restituisce 3?


- A : floor(a)
- B: float(int(a))
- C : int(10 // a)
- D : Nessuna delle opzioni sopra

## Controllare i tipi di dato

Ci sono due modi principali per verificare il tipo di dato di un oggetto. Il più semplice è usare **type**, come abbiamo visto all’inizio di questo notebook:

In [2]:
print(type(4524))
print(type('a very good dog'))

<class 'int'>
<class 'str'>


Usando questo metodo, possiamo verificare se una variabile ha un certo tipo:

In [20]:
type([3, [], {'foo': 'bar'}]) == dict

False

Un modo più “Pythonic” per farlo è utilizzare la funzione incorporata **.isinstance:**

In [22]:
isinstance('49.890', str) # Verificare se ‘49.890’ è una stringa

True

Il principale vantaggio di questa funzione rispetto a **type** è che possiamo usarla per effettuare verifiche multiple:

In [23]:
age_inputs = [11, 14, 13, '12', 12, [16]]

for age in age_inputs:
    if not isinstance(age, (int, str) ):
        print(f'{age} is {type(age)}. It should be either integer or string')

[16] is <class 'list'>. It should be either integer or string
